codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt

```diff
@@ -605,13 +611,164 @@ private class ServerHttpProtocolImplGenerator(
                         }
                         """.trimIndent(),
                         "deserializer" to deserializer,
-                        "E" to errorShape,
                     )
                 }
             }
         }
     }
 
+    // The `httpQueryParams` trait can be applied to structure members that target:
+    //     * a map of string,
+    //     * a map of list of string; or
+    //     * a map of set of string.
+    enum class QueryParamsTargetMapValueType {
+        STRING, LIST, SET;
+
+        fun asRustType(): RustType =
+            when (this) {
+                STRING -> RustType.String
+                LIST -> RustType.Vec(RustType.String)
+                SET -> RustType.HashSet(RustType.String)
+            }
+    }
+
+    private fun queryParamsTargetMapValueType(targetMapValue: Shape): QueryParamsTargetMapValueType =
+        if (targetMapValue.isStringShape) {
+            QueryParamsTargetMapValueType.STRING
+        } else if (targetMapValue.isListShape) {
+            QueryParamsTargetMapValueType.LIST
+        } else if (targetMapValue.isSetShape) {
+            QueryParamsTargetMapValueType.SET
+        } else {
+            throw ExpectationNotMetException("""
+                @httpQueryParams trait applied to non-supported target
+                $targetMapValue of type ${targetMapValue.type}
+                """.trimIndent(),
+                targetMapValue.sourceLocation)
+        }
```

Comment:
Probably just style, but I like more `when` than chained if / else if / else. Not a blocker.